cpp/include/raft/spatial/knn/detail/ivf_pq_search.cuh

```diff
@@ -873,128 +986,255 @@ struct ivfpq_compute_similarity {
    *    whether use the fused calculate+select or just calculate the distances for each
    *    query and probed cluster.
    *
+   * @param locality_hint
+   *    beyond this limit do not consider increasing the number of active blocks per SM
+   *    would improve locality anymore.
    */
   static inline auto select(bool manage_local_topk,
+                            int locality_hint,
+                            double preferred_shmem_carveout,
                             uint32_t pq_bits,
                             uint32_t pq_dim,
-                            uint32_t rot_dim,
-                            uint32_t preferred_thread_block_size,
+                            uint32_t precomp_data_count,
                             uint32_t n_queries,
                             uint32_t n_probes,
                             uint32_t topk) -> selected
   {
-    using conf_fast        = configured<true, true>;
-    using conf_no_basediff = configured<false, true>;
-    using conf_no_smem_lut = configured<true, false>;
-
-    kernel_t kernel_fast = conf_fast::kernel(pq_bits, pq_dim, manage_local_topk ? topk : 0u);
-    kernel_t kernel_no_basediff =
-      conf_no_basediff::kernel(pq_bits, pq_dim, manage_local_topk ? topk : 0u);
-    kernel_t kernel_no_smem_lut =
-      conf_no_smem_lut::kernel(pq_bits, pq_dim, manage_local_topk ? topk : 0u);
-
-    const size_t smem_threshold = 48 * 1024;
-    size_t smem_size            = sizeof(LutT) * (pq_dim << pq_bits);
-    size_t smem_size_base_diff  = sizeof(float) * rot_dim;
-
-    uint32_t n_blocks  = n_queries * n_probes;
-    uint32_t n_threads = 1024;
-    // preferred_thread_block_size == 0 means using auto thread block size calculation mode
-    if (preferred_thread_block_size == 0) {
-      const uint32_t thread_min = 256;
+    cudaDeviceProp dev_props;
+    {
       int cur_dev;
-      cudaDeviceProp dev_props;
       RAFT_CUDA_TRY(cudaGetDevice(&cur_dev));
       RAFT_CUDA_TRY(cudaGetDeviceProperties(&dev_props, cur_dev));
```

Comment:
I'm using a few different properties repeatedly in there, so I'm not sure what's faster.

I'm also thinking maybe it makes sense to add the `cudaDeviceProp` struct for the selected devices to the raft handle for lazy loading as we're doing with cublas/cusolver and other things?.. @cjnolet do you think it makes sense to do that?

Reply:
@tfeher i was just about to reference that finding made by Mark H and Thejaswi about the perf hit for the device properties. I don't think it's a bad idea to add the properties to the handle if they really can be lazy loaded. One thing to consider, though, is that we're soon going to be refactoring the handle so that the base implementation becomes more of a generic key/value store for resources. Eventually, we will have separate "handlers" that get delegated the task of loading the resources (and storing them) when requested. Shouldn't change much here but I wanted to note that the change will be coming up soon (23.02 or 23.04 timeframe) as we begin to abstract our APIs for potentially supporting different execution environments in the future (e.g. Grace).

Reply:
I definitely think we should either move the device properties or use get attribute, though. From what I recall, the loading the device properties added a significant perf hit.

Reply:
It does not significantly affect the performance of the ivf-pq search in my benchmarks though (bigann-10m, bigann-100m, glove-100). I've only noticed it takes some measurable percentage (~5-10%) of the search time (visually via nsys timeline) on very small work sizes (n_queries * n_probes).

I'd like to change this to lazy load the properties via handle, but I feel this PR has grown too much already. @cjnolet, shall do that in a follow-up?

Reply:
Sure I'm okay w/ doing it as a follow-on. Can we capture that in an issue somewhere and reference it in the PR? That way we won't lose sight of it just in case other things take priority

Reply:
Created a task: https://github.com/rapidsai/raft/issues/711 (10)